### DISCUSSION TITLE
Allow trailing semicolons at the end of property declarations

### DISCUSSION THREAD

**Discussion by irvnriir** [dotnet/csharplang#8195] [Language Ideas]
Can [CS1597](https://learn.microsoft.com/en-us/dotnet/csharp/misc/cs1597) be removed? Why saving space of 1 per hundreds of chars in expense of simplicity and even consistency, is enforced?

Is there any functional reason for it?

When writing alot, or prototyping, i get builds failing because of this. And as far i see, it has no functional purpose.

Even if it isn't needed for the lang, it could be added as a warning, but why in a world its a compilation error so can't even be disabled user-side?

**Comment by CyrusNajmabadi** [comment]
The build doesn't crash here. An error is given stating you have a superfluous token. 

We'd need to update the grammar here to allow this, and that would need a compelling argument as to *why* to allow this.

**Reply by irvnriir** [reply]
There've been a long discussion on Discord. As the last bits, the reasoning for not doing this, was that statements which affect a `{<<..>>}` block could be added after that block as a feature in future.

p.s.

this your response above, is just a request to write-up the process of how and why the following happens:
> When writing alot, or prototyping, i get builds failing because of this.

i would suggest/request to try making ~~functional~~ constructive responses next time. ty for attention.

**Reply by CyrusNajmabadi** [reply]
> i would suggest/request to try making functional responses next time.

You're asking for a change to the language, as i stated:  that would need a *compelling* argument as to why to allow this.

That still remains the case.  We won't make any changes without strong reason to do so.

> i would suggest/request to try making functional responses next time. 

As per the prior conversation, please abide by the .net code of conduct.  This includes treating people with respect and having constructive conversations on the myriad aspects of language design that we discuss here.  Thanks.

**Reply by FaustVX** [reply]
@irvnriir 

> There've been a long discussion on Discord. As the last bits, the reasoning for not doing this, was that statements which affect a `{<<..>>}` block could be added after that block as a feature in future.
> 
> p.s.
> 
> this your response is just a request to write-up the process of how and why the following happens.
> 
> > When writing alot, or prototyping, i get builds failing because of this.
> 
> i would suggest/request to try making ~functional~ constructive responses next time. ty for attention.

Maybe I'm missing something, but why are you responding to yourself ?

**Reply by irvnriir** [reply]
@FaustVX i'm not. i initially had a sentence after the ".", idk if thats allowed in English, but that doesn't cause any confusion afaik. i also was too tired to notice the ".", though in accordance to my previous sentence, i think it didn't require fixing.

Edit: the missing `above` was technically an error, but i omitted it in attempt to be more polite/neutral.

**Comment by CyrusNajmabadi** [comment]
> Is there any functional reason for it?

yes.  the grammar defines teh syntactic rules of the language.  And having our compiler not deviate from those rules is something we consider very important.  This ensures consistency among users about how to read/write C#, and it means we don't introduce deviations into our compiler that then potentially interfere with future language changes.

> When writing alot, or prototyping, i get builds failing because of this. 

Deleting hte extra semicolon seems pretty simple and cheap to do.  What is causing the extra semicolons in the first place?  What other extraneous tokens do you have when prototyping?

>  but why in a world its a compilation error so can't even be disabled user-side?

It's illegal syntax.  Like with all extra tokens or missing tokens, we can guess what you might have meant, but we really have no idea.  And since we don't know, we fail (like every other mainstream compiler out there).

**Reply by irvnriir** [reply]
> And having our compiler not deviate from those rules is something we consider very important.

to be clear. this suggestion encompasses all relevant specifications and implementations (ov), its not limited to the compiler.

> What is causing the extra semicolons in the first place? What other extraneous tokens do you have when prototyping?

reason is that in C# (most important), and even more is some other languages (also including CSV and such other), `;` serves as an ending of a statement. fact that ";" is forbidden after the blocks, has to be explicitly remembered, that distracts and doesn't allow to squeeze the spent attention into actual work.

> It's illegal syntax.

point of the topic is about changing parsing, and if needed, specifications.

**Reply by CyrusNajmabadi** [reply]
> ; serves as an ending of a statement.

These are explicitly *not* statements though.  You cannot use general members as statements.  Constructors, properties, indexers, events, operators, destructors, etc, are all not legal as statements.

>  has to be explicitly remembered,

That's true with every type of node that is not another type of node.  For example, some nodes allow blocks, others do not.  but we're not likely to allow you to go add a block to every other construct just because some of them have htem.

> > ; serves as an ending of a statement.

FWIW, this is also not true.  Many of our statements do not have semicolons.  For example, none of the loops end with a semicolon.  Neither do the 'block/checked/fixed/if/lock/switch/try/unsafe/using' statements end with a semicolon.

it's a narrower set of statements that actually end with a semicolon in our language.

**Reply by CyrusNajmabadi** [reply]
> point of the topic is about changing parsing

You asked why the compiler is giving an error and not a warning.  And i'm explaining the reasoning behind why the compiler does what it does.  Currently, our position with teh parser is that it should not deviate from the lang specification.  We have tha tposition because there have been points in the past where it *has*, and we are not happy about htat.  It means we effectively now have two specifications.  What the language itself says is legal/illegal, and what the compiler actually does.  We then have to figuer that all out, and document it, and attempt to maintain it.  Deviations between the two are undeseriable.

In terms of making hte language itself support this case.  That's possible.  Though it runs into the issue i outline below.  Namely that we are not likely to do such a change *unless* it had widespread benefit.  We make syntactic affordances like this when we think teh value is super high.  We do not when the value is not there.  It's about as simple as that.

**Reply by irvnriir** [reply]
> These are explicitly not statements though.

"statement" in common English. i would write "Statement" if meant the definition in C#.

> That's true with every type of node

i don't see how that is related to the case.

> this is also not true

i didn't mean every, even if would mean Statements.

**Reply by CyrusNajmabadi** [reply]
> "statement" in common English.

Theses are not statements in common english either.  And common english would want superflous semicolons either at the ends here either.

Furthermore, we aren't designing our language based on english.  That's not a goal for us.  

> i don't see how that is related to the case.

Because you're stating that a reason to allow this is because people have to remember they're not allowed to do this.  But that's true of the entire language.  If a person comes and says "i want to place `_` after a member, because otherwise i have to remember that i'm not allowed to do that", that's not a compelling case to make.

> i didn't mean every, even if would mean Statements.

Then i don't get yoru argument.  We use semicolons in a few places.  You're asking us to expand that to more cases.  that would again need a compelling argument as to the benefit gained by doing that.  We also use many other tokens in many other parts of the language.  But we don't have any goal/intent of allowing superfluous instances of those to show up in arbitrary locations.

As per the below, i outlined the sort of case that *can* be made for specific tokens in specific conditions.  And that case demonstrates at least *why* the value is there, and *how* it can impact a large enough group for consideration.  To be considered, this case would need a similar argument made for it.

**Reply by irvnriir** [reply]
that subject is off-topic, but function, Accessor, class and etc are all statements.
![image](https://github.com/dotnet/csharplang/assets/13949704/1fcd0925-7d01-4834-a092-b75fe39b2b8c)
* sources: [*](https://dictionary.cambridge.org/dictionary/english/statement?q=Statement), [*](https://en.wikipedia.org/wiki/Statement_(logic)).

> Because you're stating that a reason to allow this

they have to remember because of it's specific usage situation and specific subject location, not because they randomly decided to.

> Then i don't get yoru argument.

not every part of syntax lays down into patterns to cause noticeable or any effects on other parts. im explaining how and why. whether it affects one user or other, and how much, is covered by each's own understanding, and the impact topic.

**Reply by CyrusNajmabadi** [reply]
> that subject is off-topic, but function, Accessor, class and etc are all statements.

Then everything in c# is a statement.  Because everything expresses an idea of some form. 

And we are certainly not going to allow superfluous semicolons after all constructs in the language just because someone vaguely thinks of all constructs as statements.

**Reply by CyrusNajmabadi** [reply]
> they have to remember because of it's specific usage situation and specific subject location, not because they randomly decided to.

Yes.  And that applies to any user for any syntactic construct in any location.   You stated, explicitly: 

>  fact that ";" is forbidden after the blocks, has to be explicitly remembered

But htat's literally true of *everything* in the language.  Practically every token is not actually legal at the end of nearly all constructs.  I can't put superfluous periods at the end of things (or in the middle of many constructs) despite being legal in others.  I can't put random curlies wherever i might want, even though they're legal at the end of a bunch of constructs.  This is hte norm for *all* programming languages.  You can't just assume or expect that some tokens that are legal in some places will be legal in unrelated places.  And yes, i guess it's the case you have to remember/know that.  That's part of what learning a language is about. 

You've already done it for an absolutely huge number of other cases.  It's not at all clear why semicolon-after-property isn't just one more case you just learn and remember from that point on.  given no other feedback in 25 years from anyone else having an issue here... it seems like there really isn't any sort of problem people are running into.  they know they can't write `;` here just like they can't write it in the thousands of other locations in the lang where it isn't legal.  :)

**Reply by irvnriir** [reply]
> Then everything in c# is a statement. Because everything expresses an idea of some form.

yes

> Yes. And that applies to any user for any syntactic construct in any location.

i meant that not every possible activity has a reason, there is no need to remember to not do something that doesn't have reason to be done.

> But that's literally true of everything in the language.

in addition to the above, the subject places are related.

> it seems like there really isn't any sort of problem people are running into.

for the most part because they adjust, not because its the same as some other cases.

**Reply by CyrusNajmabadi** [reply]
> i meant that not every possible activity has a reason, there is no need to remember to not do something that doesn't have reason to be done.

That applies to your case as well.  There's no reason to place a random semicolon in an arbitrary position like this.  Absent a reason to do it, we don't allow these things.  Since there's no reason to do this, there's no need to remember anything about this (as per your direct argument above) :)

> in addition to the above, the subject places are related.

There's no relationship between a property and any of the other constructs that use semicolons. 

> for the most part because they adjust

If they've adjusted and have no issue, then there's no need to change anything here.  We invest the efforts where people continue having problems.

**Reply by irvnriir** [reply]
there is a reason, which i described. regardless of whether you allow to do it or not :) . i won't harm: my analytical and processing qualities, just for adjusting to this. i have to remember the specific case.

> There's no relationship between

it depends on mental structures, there is functional relationship.

> If they've adjusted and have no issue

it worths mentioning, and i did. also, various things can unexpectedly happen to need little effort for implementation, or have many users being affected but not mentioning it.

**Reply by CyrusNajmabadi** [reply]
> it depends on mental structures, there is functional relationship.

Anyone can make the claim about any language concepts. The cases I listed before that you said:

> there is no need to remember to not do something that doesn't have reason to be done.

Could have someone say that in their mental structure there is a relationship. 

We can't design things around you having made a particular connection like this. Imagine if every user did what you did, just about a different part of the language. I've, you've already imagined that. You remember the other cases in your prior responses. Now imagine any of those users seeing your case and rejecting it for the same reason. 

This is why we can't do a one off feature for a single user. We need the use case to be compelling and widespread to be valuable to a large portion of the user base.

**Reply by irvnriir** [reply]
> Could have someone say that in their mental structure there is a relationship.

i don't argue that. i explained the one here for the most part.

ik that everything in such cases can be proven or disproven as reasonable, but i won't go into that here. thats why i mentioned in Discord that having counter arguments is important, i wouldn't care about any connection anyone makes, if: it makes sense, or supporting it doesn't cause any issues (not implying that this case doesn't).

**Reply by CyrusNajmabadi** [reply]
> having counter arguments is important

The counter arguments (which were already shared in discord earlier) are simply that there doesn't seem to be value here. It adds complexity. And it potentially interferes with future design space.  

Furthermore, nothing had actually been presented why this even makes sense for the language, beyond vague claims that this relates to English sentences somehow, and that your mental view is that this should work and you don't like having to remember that it doesn't. 

To be worthwhile, this would have to demonstrate substantial value just to get traction. 

As a point of comparison, trailing commas have substantial value, but the proposals on that still have not moved forward. And that's a proposal with lots of people asking for it, and clear *demonstrated* cases where it makes life nicer for many users during normal coding patterns.

**Reply by irvnriir** [reply]
doesn't it already work? (C# 5)
![image](https://github.com/dotnet/csharplang/assets/13949704/50c4c4c2-71e1-4b9b-b933-0faef26c3661)

just little fix: there is no complexity.

**Reply by CyrusNajmabadi** [reply]
It works in some constructs. But there are many more it does not.  We have examples in discord with you :-)

**Reply by CyrusNajmabadi** [reply]
> there is no complexity

There is. We shared a huge area of complexity with you in discord. This overlaps future design space ideas we have.  

This is a critical part of lang design.  The designs have to fit with the past and also feel like they will not impede the future.  

Furthermore, complexity (even if small) would be only one part of the concerns here.  The other concerns already listed *still* apply. All of that needs to be considered *as a whole* for every proposal.

**Reply by irvnriir** [reply]
> adds complexity. And it potentially interferes with future design space.

its 2 different things. whether to allow "};" or not, doesn't affect complexity, though that depends on whether allowing "};" makes sense to you. in code its 1 `if` in a list, so no structural complexity either.

> Furthermore

thats why i said "little fix".

**Reply by CyrusNajmabadi** [reply]
It does affect complexity. We went over this in depth in discord.  I also listed the issue above with this, especially as it impacts future design work in this space.

**Reply by CyrusNajmabadi** [reply]
> in code its 1 if in a list, so no structural complexity either.

This is not the case. Every feature that operates syntactically on property declarations needs to be updated.  That is inherent complexity. 

Note. This is not a hypothetical.  We've slightly changed the syntax to support tiny changes like this to a few constructs.  Like allowing `using` to end with a semicolon, as well as namespace decls.  As well as a ton of other "seemingly tiny" grammar changes. They have a lot of complexity because we have thousands and thousands of lines that expect the tree to look a certain way.  

As a core example, this will break intelligence for the token following the ; until we go and fix that feature up to understand this.  That's just one of a huge number of features.

**Comment by CyrusNajmabadi** [comment]
> Why saving space of 1 per hundreds of chars in expense of simplicity and even consistency, is enforced?

It's not to 'save space'.  It is simply that we have grammar rules.  And if one wants those grammar rules to encompass more token sequences, then the grammar needs to be updated accordingly.  The spec also needs to then say what the impact of that extra token is (even to say that it is ignored).

the reason we don't *normally* do this is simply that it adds excess to the language, compiler, and all higher tooling that in practice doesn't make things substantively better for users.  We generally only want this sort of syntactic flexibility if there is a big win for users.  For example, there are compelling cases to be made about allowing trailing commas in lists, due to users have something like:

```c#
   x1, 
   x2, 
   x3, 
   x4,
```

And wanting to be able to reorder those easily by just moving lines around.  If the last comma is not allowed, this becomes more annoying.  This is at least something we can identify as something a large set of users do and find value in, and so we allow this sort of extra, technically not necessary token.

However, there an *infinite* number of extraneous tokens people could place anywhere.  adding support for everything is not viable.  So we add hte support where there is a compelling case for it, and there is a clear indication a lot of people would be helped out.  That is not the case here afaict.

**Reply by irvnriir** [reply]
the topic is about a specific case, not randomly picked token at randomly picked spot. but at the time i don't state this to have a large or widespread impact.

**Reply by CyrusNajmabadi** [reply]
> the topic is about a specific case

yes.  And i'm explaining how we assess individual cases.  That involves showing an example of where we would consider adding support for extra syntax, and the difference between the two. 

> but at the time i don't state this to have a large or widespread impact.

Right. And i'm stating that that's something we care about.  We are unlikely to make changes if they're not goign to have substantive benefit to our ecosystem. There are millions of users we have to cater to, and we cannot invest in one-off features that will help one user out in their particular use case.

**Reply by irvnriir** [reply]
> .. That involves showing ..

your example, and parts of the response, are good. i should've quote the part which i was responding too.

**Comment by CyrusNajmabadi** [comment]
I'm locking this at this point.  Both in discord and here there appears to be no attempt to provide any sort of justification for this feature. Furthermore, the op doesn't seem interested in understanding the design process or the critiques levied against this.  

This conversation seems to have run its course and no efforts are being made to substantively engage here. 

More time has been spent arguing there and here than the ecosystem would ever will back in time spent on this proposal. @irvnriir I recommend for future proposals that you be more willing to accept that a feature needs to be more valuable to us than just helping you so you can avoid a single char mistake. Think broadly about the millions of other users out there and whether or not this will meaningfully affect them.  If so, be prepared to argue and demonstrate why that is.
